BedrockConflictManager.cpp

```diff
@@ -64,3 +64,73 @@ string BedrockConflictManager::generateReport() {
     }
     return out.str();
 }
+
+mutex PageLockGuard::controlMutex;
+map<int64_t, mutex> PageLockGuard::mutexes;
+list<int64_t> PageLockGuard::mutexOrder;
+map<int64_t, list<int64_t>::iterator> PageLockGuard::mutexOrderFastLookup;
+map<int64_t, int64_t> PageLockGuard::mutexCounts;
+
+PageLockGuard::PageLockGuard(int64_t page) : _page(page) {
+    if (_page == 0) {
+        return;
+    }
+
+    // We need access to the mutex outside of the control lock, so that we aren't blocking other PageLockGuard users while we wait for the page.
+    mutex* m;
+    {
+        lock_guard<mutex> lock(controlMutex);
+        auto mutexPair = mutexes.find(_page);
+        if (mutexPair == mutexes.end()) {
+            mutexPair = mutexes.emplace(piecewise_construct, forward_as_tuple(_page), forward_as_tuple()).first;
+            mutexCounts.emplace(make_pair(_page, 1l));
+            mutexOrder.push_front(_page);
+            mutexOrderFastLookup.emplace(make_pair(_page, mutexOrder.begin()));
+        } else {
+            // Increment the reference count.
+            mutexCounts[_page]++;
+
+            // If the current mutex was already at the front of the order list, no updates are needed.
+            if (mutexOrder.front() != _page) {
+                // Erase the old location of this mutex in the order list and move it to the front.
+                mutexOrder.erase(mutexOrderFastLookup.at(_page));
+                mutexOrder.push_front(_page);
+
+                // And save the new fast lookup at the front.
+                mutexOrderFastLookup[_page] = mutexOrder.begin();
+            }
+        }
+
+        static const size_t MAX_PAGE_MUTEXES = 500;
```

Comment:
Was this chosen arbitrarily? What if we make it 1000 instead?

Reply:
Yes. If we do that, then it will be 1000.

Reply:
Let's make it configurable so we can play with it live without needing to deploy?

Reply:
I don't think this will make any difference. To have more than 500 mutexes in use requires us to have at least 500 conflicting commands committing simultaneously, and for the mutexes to actually do anything, there needs to be multiple commands using each one, so practically we need 1000+ commands executing simultaneously. Also, the mutexes can't actually be pruned while they're in use, so if there are 1000 simultaneous conflicts on different pages, the list will just expand until some aren't being used anymore.

Reply:
In that case, will you add all that as a comment above this line?